Sources/WebSocketKit/PMCE.swift

```diff
@@ -0,0 +1,753 @@
+        if logging {
+            logger.trace("PMCE: decompressing  \(startSize) bytes for \(frame.opcode)")
+        }
+        logger.trace("PMCE: config: \(serverConfig) \(clientConfig)")
+        
+        if takeover {
+            data = pad(buffer:data)
+        }
+
+        let decompressed =
+        try data.decompressStream(with: self.decompressorBox.value!,
+                                  maxSize: .max,
+                                  allocator: channel.allocator)
+        if logging {
+            let endTime = Date()
+            
+            let endSize = decompressed.readableBytes
+            
+            logger.trace("deompressed \(startSize) to \(endSize) bytes @ \(endSize/startSize) ratio from")
+            logger.trace(" \(serverConfig.deflateConfig.zlibConfig) \(clientConfig.deflateConfig.zlibConfig)")
+
+            
+            logger.trace("in \(startTime.distance(to: endTime))")
+        }
+        
+        if !takeover {
+            if logging { logger.trace("PMCE: resetting decompressoer stream.") }
+            try decompressorBox.value?.resetStream()
+        }else {
+            if logging { logger.trace("PMCE: not restting decompressor stream.")}
+        }
+        
+        let newFrame = WebSocketFrame(fin: frame.fin,
+                                      rsv1: false,
+                                      rsv2: frame.rsv2,
+                                      rsv3: frame.rsv3,
+                                      opcode: frame.opcode,
+                                      maskKey: frame.maskKey,
+                                      data: decompressed,
+                                      extensionData: nil)
+        return newFrame
+    }
+
+    /// Server decomp uses this as RFC-7692 says client must mask msgs but server must not.
+    public func unmasked(frame maskedFrame: WebSocketFrame) -> WebSocketFrame {
+        logger.trace("PMCE: unmasking \(maskedFrame)")
+        guard let key = maskedFrame.maskKey else {
+            logger.trace("PMCE: tried to unmask a frame that isnt already masked.")
+            return maskedFrame
+        }
+        
+        var unmaskedData = maskedFrame.data
+        unmaskedData.webSocketUnmask(key)
+        return WebSocketFrame(fin: maskedFrame.fin,
+                              rsv1: maskedFrame.rsv1,
+                              rsv2: maskedFrame.rsv2,
+                              rsv3: maskedFrame.rsv3,
+                              opcode: maskedFrame.opcode,
+                              maskKey: nil,
+                              data: unmaskedData,
+                              extensionData: maskedFrame.extensionData)
+    }
+
+    ///
+    private let logger = Logger(label: "PMCE")
+    private let _logging:NIOLockedValueBox<Bool>
+    private let _enabled:NIOLockedValueBox<Bool>
+
+    // Converts windowBits to size of window.
+    private static func sizeFor(bits:UInt8) -> Int32 {
+        2^Int32(bits)
```

Comment:
There's no possible way this makes sense:
```swift
  1> 2^15
$R0: Int = 13
```
`^` is XOR, not exponentiation. You want this:
```suggestion
        1 << Int32(bits)
```